mgmt/utils/MgmtMarshall.h

```diff
@@ -61,17 +75,112 @@ struct MgmtMarshallData {
+ssize_t mgmt_message_read(int fd); // to allow for expansion
+
+template <typename T, typename... Rest>
+ssize_t
+mgmt_message_read(int fd, T first, Rest &&... rest)
+{
+  ssize_t nfirst = mgmt_message_read(fd, first);
+  ssize_t nrest  = mgmt_message_read(fd, std::forward<Rest>(rest)...);
+  return (nfirst != -1 && nrest != -1) ? nfirst + nrest : -1;
+}
+/// mgmt_message_read ---------------------------------------------------------
+
+/// mgmt_message_write --------------------------------------------------------
+ssize_t mgmt_message_write(int fd, MgmtMarshallInt *field);
+ssize_t mgmt_message_write(int fd, MgmtMarshallLong *field);
+ssize_t mgmt_message_write(int fd, MgmtMarshallString *field);
+ssize_t mgmt_message_write(int fd, MgmtMarshallData *field);
+ssize_t mgmt_message_write(int fd); // to allow for expansion
+
+template <typename T, typename... Rest>
+ssize_t
+mgmt_message_write(int fd, T first, Rest &&... rest)
+{
+  ssize_t nfirst = mgmt_message_write(fd, first);
+  ssize_t nrest  = mgmt_message_write(fd, std::forward<Rest>(rest)...);
+  return (nfirst != -1 && nrest != -1) ? nfirst + nrest : -1;
+}
+/// end mgmt_message_write ----------------------------------------------------
+
+/// mgmt_message_marshall -----------------------------------------------------
+ssize_t mgmt_message_marshall(void *buf, size_t remain, MgmtMarshallInt *field);
+ssize_t mgmt_message_marshall(void *buf, size_t remain, MgmtMarshallLong *field);
+ssize_t mgmt_message_marshall(void *buf, size_t remain, MgmtMarshallString *field);
+ssize_t mgmt_message_marshall(void *buf, size_t remain, MgmtMarshallData *field);
+ssize_t mgmt_message_marshall(void *buf, size_t remain); // to allow for expansion
+
+template <typename T, typename... Rest>
+ssize_t
+mgmt_message_marshall(void *buf, size_t remain, T first, Rest &&... rest)
+{
+  if (buf == nullptr) {
+    return -1;
+  }
+  ssize_t nfirst = mgmt_message_marshall(buf, remain, first);
+  ssize_t nrest  = mgmt_message_marshall(static_cast<uint8_t *>(buf) + nfirst, remain - nfirst, std::forward<Rest>(rest)...);
+  return (nfirst != -1 && nrest != -1) ? nfirst + nrest : -1;
+}
+/// end mgmt_message_marshall ---------------------------------------------------
+
+/// mgmt_message_parse ----------------------------------------------------------
+ssize_t mgmt_message_parse(const void *buf, size_t len, MgmtMarshallInt *field);
+ssize_t mgmt_message_parse(const void *buf, size_t len, MgmtMarshallLong *field);
+ssize_t mgmt_message_parse(const void *buf, size_t len, MgmtMarshallString *field);
+ssize_t mgmt_message_parse(const void *buf, size_t len, MgmtMarshallData *field);
+ssize_t mgmt_message_parse(const void *buf, size_t len); // to allow for expansion
+
+template <typename T, typename... Rest>
+ssize_t
+mgmt_message_parse(const void *buf, size_t len, T first, Rest &&... rest)
+{
+  if (buf == nullptr) {
+    return -1;
+  }
+  ssize_t nfirst = mgmt_message_parse(buf, len, first);
```

Comment:
Correct but for performance there's no point in doing the second parse if the first doesn't work. E.g. something more like
```
ssize_t n1, n2;
return -1 != (n1 = mgmt_message_parse(buf, len, first)) &&
    -1 != mgmt_message_parse(buff + n1, len -n1, std::forward<Rest>(rest)...)) ? n1 + n2 : -1;
```

Reply:
updated.